Replace URI.escape/unescape

kbruccoleri opened the pull request:
These methods are removed from Ruby 3.
We are using the URI::RFC2396_Parser to maintain current behavior.

kbruccoleri committed:
Replace URI.escape/unescape

These methods are removed from Ruby 3.
We are using the URI::RFC2396_Parser to maintain current behavior.

MarkyMarkMcDonald reviewed (commented):
Looks like the new parser has a decent amount of logic / stuff going on in the initializer https://github.com/ruby/uri/blob/master/lib/uri/rfc2396_parser.rb#L338

are any of these code paths hotspots that we should memoize the parser for instead of reconstructing every time?

kbruccoleri commented:
> Looks like the new parser has a decent amount of logic / stuff going on in the initializer https://github.com/ruby/uri/blob/master/lib/uri/rfc2396_parser.rb#L338
> 
> are any of these code paths hotspots that we should memoize the parser for instead of reconstructing every time?

That's a good call out, thanks for finding that. I was wondering that, but was staying on the side of caution to prevent introducing any tricky bugs. I can definitely take another look and see what we can do.

MarkyMarkMcDonald reviewed (approved):
I looked through this again and can't think of any hot spots so I'm down to ship. But I don't see any state held in the parser that's mutated by `#escape` and we're already calling the same parser multiple times; I could see making our own module that memoizes the parser globally and provides a static method for usage (~maybe `StringHelper#escape_url`?~ forgot we're not in GHR >.<).

kbruccoleri committed:
Opt to use URI::DEFAULT_PARSER

Which is a singleton instance of the URI::RFC2396_Parser
https://github.com/ruby/uri/blob/4346daac75b6b7772e67c81b1e0eab80c4f87034/lib/uri/common.rb#L19
https://github.com/ruby/uri/blob/4346daac75b6b7772e67c81b1e0eab80c4f87034/lib/uri/common.rb#L24
